test/integration/utility.cc

```diff
@@ -75,27 +83,104 @@ void BufferingStreamDecoder::onResetStream(Http::StreamResetReason, absl::string
   ADD_FAILURE();
 }
 
+struct ConnectionCallbacks : public Network::ConnectionCallbacks {
+  ConnectionCallbacks(Event::Dispatcher& dispatcher) : dispatcher_(dispatcher) {}
+
+  // Network::ConnectionCallbacks
+  void onEvent(Network::ConnectionEvent event) override {
+    if (event == Network::ConnectionEvent::Connected) {
+      connected_ = true;
+      dispatcher_.exit();
+    } else if (event == Network::ConnectionEvent::RemoteClose) {
+      dispatcher_.exit();
+    } else {
+      if (!connected_) {
+        // Before handshake gets established, any connection failure should exit the loop. I.e. a
+        // QUIC connection may fail of INVALID_VERSION if both this client doesn't support any of
+        // the versions the server advertised before handshake established. In this case the
+        // connection is closed locally and this is in a blocking event loop.
+        dispatcher_.exit();
+      }
+    }
+  }
+
+  void onAboveWriteBufferHighWatermark() override {}
+  void onBelowWriteBufferLowWatermark() override {}
+
+  Event::Dispatcher& dispatcher_;
+  bool connected_{false};
+};
+
+Network::TransportSocketFactoryPtr IntegrationUtil::createQuicClientTransportSocketFactory(
+    Server::Configuration::TransportSocketFactoryContext& context,
+    const std::string& san_to_match) {
+  envoy::extensions::transport_sockets::quic::v3::QuicUpstreamTransport
+      quic_transport_socket_config;
+  auto* tls_context = quic_transport_socket_config.mutable_upstream_tls_context();
+  initializeUpstreamTlsContextConfig(
+      Ssl::ClientSslTransportOptions().setAlpn(true).setSan(san_to_match).setSni("lyft.com"),
+      *tls_context);
+
+  envoy::config::core::v3::TransportSocket message;
+  message.mutable_typed_config()->PackFrom(quic_transport_socket_config);
+  auto& config_factory = Config::Utility::getAndCheckFactory<
+      Server::Configuration::UpstreamTransportSocketConfigFactory>(message);
+  return config_factory.createTransportSocketFactory(quic_transport_socket_config, context);
+}
+
 BufferingStreamDecoderPtr
 IntegrationUtil::makeSingleRequest(const Network::Address::InstanceConstSharedPtr& addr,
                                    const std::string& method, const std::string& url,
                                    const std::string& body, Http::CodecClient::Type type,
                                    const std::string& host, const std::string& content_type) {
-
   NiceMock<Stats::MockIsolatedStatsStore> mock_stats_store;
   NiceMock<Random::MockRandomGenerator> random;
   Event::GlobalTimeSystem time_system;
   NiceMock<Random::MockRandomGenerator> random_generator;
   Api::Impl api(Thread::threadFactoryForTest(), mock_stats_store, time_system,
                 Filesystem::fileSystemForTest(), random_generator);
   Event::DispatcherPtr dispatcher(api.allocateDispatcher("test_thread"));
+  ConnectionCallbacks connection_callbacks(*dispatcher);
```

Comment:
this function has gotten a lot longer and harder to follow due to the handling of h3 in it. Can we think of ways to refactor this so that the function structure remains more like it used to be?

E.g. one option would be to just add an early dispatch:
```
  if (h3 mode) {
    return makeSingleH3Request(...);
  }
```
and then factor out the common blocks from the h3 and non-h3 impls into some helpers.

WDYT?

Reply:
Thanks for suggesting the refactoring! Surprisingly, this fix the TSAN issue I asked earlier. I found that createQuicUpstreamTransportSocketFactory() which was also called under h2 and h1 earlier somehow breaks the stats histogram Expectation. But I still don't understand how could a client util function lead to server stats change. This function calls a registered factory to create a quic upstream transport socket factory for the test client.

Reply:
Yeah the reason is that went looking at the code in its previous state, it was very heard to discern whether the behavior was the same in H2 with/without your change. I suspect it wasn't, and that was what was affecting the integration test tsan.

Reply:
I'd guess this is inducing the client to change the way it communicates with the server in a stat-observable way, but I'd have to dig in with the debugger to learn more.

Reply:
I reproduced the same failure by adding a line to print trace in a clean branch https://github.com/envoyproxy/envoy/compare/main...danzh2010:admin_test?expand=1. Is there some existing race in that test?

Reply:
Interesting. Would a sleep there also cause the test to fail? How about printing a bunch of text to stderr?

That trace macro doesn't look like it ought to wake up anything in stats to change behavior, so I suspect a sleep or simple I/O might do the same thing.